### PR TITLE
HPCC-28618 Refactor jhtree to avoid fetching entire row when only key fields needed

### PR DESCRIPTION
**Pull request by richardkchapman**
Refactor jhtree code to only use keyed portion of buffer when searching, and fetch full record only when required for delivery, in preparation for potentially compressing the keyed and unkeyed portions of the record separately.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [x] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:

Regression suite

### CODE REVIEW

File: system/jhtree/ctfile.cpp [hunk 1]
```diff
@@ -810,6 +821,33 @@ bool CJHTreeNode::contains(const char *src) const
...
 
+void CJHTreeNode::dump(FILE *out, int length, unsigned rowCount, bool raw) const
+{
+    if (!raw)
+        fprintf(out, "Node dump: fpos(%" I64F "d) type %s\n", getFpos(), getNodeTypeName());
+    if (rowCount==0 || rowCount > getNumKeys())
+        rowCount = getNumKeys();
+    for (unsigned int i=0; i<rowCount; i++)
+    {
+        char *dst = (char *) alloca(getKeyLen()+50);
```
Comment (ghalliday) [inline system/jhtree/ctfile.cpp:832]: +50?  
Also can alloca in a loop be dangerous?  Safer to dynamically allocate e.g. with MemoryAttr or similar.

File: system/jhtree/ctfile.hpp [hunk 2]
```diff
@@ -229,13 +225,16 @@ class jhtree_decl CNodeBase : public CWritableKeyNode
...
 {
 protected:
+    size32_t keyLen = 0;
+    size32_t keyCompareLen = 0;
     size32_t keyRecLen;
     char *keyBuf;
 
     void unpack(const void *node, bool needCopy);
     unsigned __int64 firstSequence;
     size32_t expandedSize;
```
Comment (ghalliday) [inline system/jhtree/ctfile.hpp:235]: unrelated.  Moving this after keyRecLen would save 8 bytes packing.

**Review by ghalliday — COMMENTED** [review]
@richardkchapman didn't spot much on a fisrst scan.  
Do you need two separate buffers though, or can you use a single buffer with the flag to indicate it is full?
Then you can call a new function appendPayloadValue(<idx>, buffer), and avoid re-copying the keyed portion.  For some indexes that are (keyed=>filepos) that would avoid the bulk of the work.  It would also avoid extra work in a new format where keyed and non keyed are compressed separately.

**Comment by richardkchapman** [comment]
I could probably combine the buffers. I didn't in the first pass because it made it easier to be sure I was not reading/writing off the end of the keyed bit (using valgrind), but also because I'm not sure whether the "full" buffer is really needed at all, or whether I will eventually be able to move the data directly into the user-supplied destination. At present it seems the only real reason for the "full" buffer is to have something that has moved the fpos to the end.

**Review by ghalliday — APPROVED** [review]

**Pull request merged**